### PR TITLE
[semantic-arc-opts] Eliminate dead trivial instructions /after/ the m…

### PR DESCRIPTION
**Pull request by gottesmm**
…ain loop to avoid issues around adding/removing things from the worklist.

rdar://55127027

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test

File: lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp [hunk 1]
```diff
@@ -251,6 +247,23 @@ bool SemanticARCOptVisitor::processWorklist() {
...
+  // be empty due to the loop above. We enforce this programatically with the
+  // assert.
+  assert(worklist.empty() && "Expected drained worklist so we don't have to "
+                             "remove dead insts form it");
+  while (!deadTrivialInsts.empty()) {
+    auto val = deadTrivialInsts.pop_back_val();
+    if (!val)
+      continue;
+    recursivelyDeleteTriviallyDeadInstructions(
+        *val, true /*force*/,
```
Comment (nate-chandler) [inline lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp]: Out of curiosity, why is `force=true` appropriate here?  Also, why was it previously necessary to add operands and results to the worklist but is no longer?

**Review by atrick — APPROVED** [review]
LGTM

File: lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp (hunk 1 above)
Comment (gottesmm) [reply lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp]: The reason why I put force here is b/c if I don't put force it just does the isTriviallyDead check again.

The reason why I removed the operand/result thing is I find with the test case in question that recursivelyDeleteTriviallyDeadInstructions can re-add the struct_extract from the test case to the worklist as a result of an instruction.

I just avoided all of the problems by doing it later.

Comment (atrick) [reply lib/SILOptimizer/Mandatory/SemanticARCOpts.cpp]: > The reason why I put force here is b/c if I don't put force it just does the isTriviallyDead check again.

Anytime we violate the spirit of the API like this there should be a comment explaining the cleverness.

**Pull request merged**